twins/mongodb/twins.go

```diff
@@ -0,0 +1,229 @@
+//
+// Copyright (c) 2019
+// Mainflux
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package mongodb
+
+import (
+	"context"
+
+	"github.com/mainflux/mainflux/twins"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+const (
+	maxNameSize            = 1024
+	twinsCollection string = "twins"
+)
+
+type twinRepository struct {
+	db *mongo.Database
+}
+
+var _ twins.TwinRepository = (*twinRepository)(nil)
+
+// NewTwinRepository instantiates a MongoDB implementation of twin
+// repository.
+func NewTwinRepository(db *mongo.Database) twins.TwinRepository {
+	return &twinRepository{
+		db: db,
+	}
+}
+
+// Save persists the twin
+func (tr *twinRepository) Save(ctx context.Context, tw twins.Twin) (string, error) {
+	if len(tw.Name) > maxNameSize {
+		return "", twins.ErrMalformedEntity
+	}
+
+	coll := tr.db.Collection(twinsCollection)
+
+	if _, err := tr.RetrieveByID(ctx, tw.ID); err == nil {
+		return "", twins.ErrConflict
+	}
+	if _, err := tr.RetrieveByKey(ctx, tw.Key); err == nil {
```

Comment:
What exactly you want to accomplish here ? To make sure that ID and Key are unique ? I don't like 2 additional queries in Save().

Reply:
Maybe you can check Indexing 
```Inserting an duplicate value for any key that is part of a unique index, such as _id, throws an exception. The following attempts to insert a document with a _id value that already exists:```